src/haz3lcore/dynamics/Transition.re

```diff
@@ -333,15 +336,18 @@ module Transition = (EV: EV_MODE) => {
         );
       Constructor;
     | Ap(dir, d1, d2) =>
+      // TODO I don't know why this needs to be final
       let. _ = otherwise(env, (d1, (d2, _)) => Ap(dir, d1, d2) |> rewrap)
       and. d1' =
-        req_value(req(state, env), d1 => Ap1(dir, d1, d2) |> wrap_ctx, d1)
+        req_final(req(state, env), d1 => Ap1(dir, d1, d2) |> wrap_ctx, d1)
```

Comment:
I still don't have a strong sense as to why this needs to be req_final but it remains indet otherwise.

Reply:
@Negabinary not sure I understand what req_value was doing here since we still need to handle casts? are casts considered values?

In any case, deferredap should be treated as if it were a lambda, so it should be a value.

Let's chat briefly about this with @7h3kk1d at the beginning of our meeting tomorrow.

Reply:
@cyrus-  Oh I see, yeah I guess if you still want to do the function cast rules when the function is indet then you do need that to be req_final

Reply:
I got all of the cases I can think of working. I'm struggling to find a case that req_final catches that req_value doesn't though. There might be something wrong with `Constructor`